services/rpcfilters/logs_filter.go

```diff
@@ -0,0 +1,120 @@
+package rpcfilters
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	ethereum "github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type logsFilter struct {
+	mu                  sync.Mutex
+	logs                []types.Log
+	lastSeenBlockNumber uint64
+	lastSeenBlockHash   common.Hash
+
+	id    rpc.ID
+	crit  ethereum.FilterQuery
+	timer *time.Timer
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	done   chan struct{}
+}
+
+func (f *logsFilter) add(data interface{}) {
+	logs := data.([]types.Log)
+	filtered, num, hash := filterLogs(logs, f.crit, f.lastSeenBlockNumber, f.lastSeenBlockHash)
+	f.mu.Lock()
+	f.lastSeenBlockNumber = num
+	f.lastSeenBlockHash = hash
+	if len(filtered) > 0 {
+		f.logs = append(f.logs, filtered...)
+	}
+	f.mu.Unlock()
+}
+
+func (f *logsFilter) pop() interface{} {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.logs == nil {
+		return []types.Log{}
```

Comment:
Why do we need to return an empty slice instead of `nil`?

Reply:
that is probably to return json array instead of null

Reply:
It should be handled in the RPC method then. Serialization logic should be in one place. Something like this:
```go
func (a *API) GetLog() []types.Log {
    logs := f.pop()
    if logs == nil {
        // to enforce rendering as an empty array
        return []types.Log{}
    }
    return logs
}
```